node/network/approval-distribution/src/lib.rs

```diff
@@ -89,6 +97,138 @@ impl RecentlyOutdated {
 	}
 }
 
+struct SessionTopology {
+	peers_x: HashSet<PeerId>,
+	validator_indices_x: HashSet<ValidatorIndex>,
+	peers_y: HashSet<PeerId>,
+	validator_indices_y: HashSet<ValidatorIndex>,
+}
+
+impl SessionTopology {
+	// Given the originator of a message, indicates the part of the topology
+	// we're meant to send the message to.
+	fn required_routing_for(&self, originator: ValidatorIndex, local: bool) -> RequiredRouting {
+		if local {
+			return RequiredRouting::GridXY
+		}
+
+		let grid_x = self.validator_indices_x.contains(&originator);
+		let grid_y = self.validator_indices_y.contains(&originator);
+
+		match (grid_x, grid_y) {
+			(false, false) => RequiredRouting::None,
+			(true, false) => RequiredRouting::GridY, // messages from X go to Y
+			(false, true) => RequiredRouting::GridX, // messages from Y go to X
+			(true, true) => RequiredRouting::GridXY, // if the grid works as expected, this shouldn't happen.
```

Comment:
I have a slight preference for `use always_assert::never` as we do in PVF subsystem.

Reply:
This thing is like a `debug_assert` in a nice wrapping. So this will panic only in CI/testnet, which is better than not noticing it at all.

Reply:
I see. Well, we could do a follow-up for that.